rspace/src/main/scala/coop/rchain/rspace/RSpace.scala

```diff
@@ -92,66 +98,65 @@ class RSpace[C, P, A, R, K](val store: IStore[C, P, A, K], val branch: Branch)(
-       * affecting the actual store contents.
-       */
-
-      val channelToIndexedData = channels.map { (c: C) =>
-        c -> Random.shuffle(store.getData(txn, Seq(c)).zipWithIndex)
-      }.toMap
-
-      val options: Option[Seq[DataCandidate[C, R]]] =
-        extractDataCandidates(channels.zip(patterns), channelToIndexedData, Nil).sequence
-
-      options match {
-        case None =>
-          store.removeAll(txn, channels)
-          store.putWaitingContinuation(
-            txn,
-            channels,
-            WaitingContinuation(patterns, continuation, persist = true, consumeRef))
-          for (channel <- channels) store.addJoin(txn, channel, channels)
-          logger.debug(s"""|consume: no data found,
+    val consumeRef = Consume.create(channels, patterns, continuation, true)
+    installs.update(_.updated(channels, Install(patterns, continuation, m)))
+
+    /*
+     * Here, we create a cache of the data at each channel as `channelToIndexedData`
+     * which is used for finding matches.  When a speculative match is found, we can
+     * remove the matching datum from the remaining data candidates in the cache.
+     *
+     * Put another way, this allows us to speculatively remove matching data without
+     * affecting the actual store contents.
+     */
+
+    val channelToIndexedData = channels.map { (c: C) =>
+      c -> Random.shuffle(store.getData(txn, Seq(c)).zipWithIndex)
+    }.toMap
+
+    val options: Option[Seq[DataCandidate[C, R]]] =
+      extractDataCandidates(channels.zip(patterns), channelToIndexedData, Nil).sequence
+
+    options match {
+      case None =>
+        store.removeAll(txn, channels)
+        store.installWaitingContinuation(
+          txn,
+          channels,
+          WaitingContinuation(patterns, continuation, persist = true, consumeRef))
+        for (channel <- channels) store.addJoin(txn, channel, channels)
+        logger.debug(s"""|consume: no data found,
                            |storing <(patterns, continuation): ($patterns, $continuation)>
                            |at <channels: $channels>""".stripMargin.replace('\n', ' '))
-          None
-        case Some(dataCandidates) =>
-          store.eventsCounter.registerInstallCommEvent()
-
-          eventLog.update(COMM(consumeRef, dataCandidates.map(_.datum.source)) +: _)
-
-          dataCandidates.foreach {
-            case DataCandidate(candidateChannel, Datum(_, persistData, _), dataIndex)
-                if !persistData =>
-              store.removeDatum(txn, Seq(candidateChannel), dataIndex)
-            case _ =>
-              ()
-          }
-          logger.debug(s"consume: data found for <patterns: $patterns> at <channels: $channels>")
-          Some((continuation, dataCandidates.map(_.datum.a)))
-      }
+        None
+      case Some(_) =>
+        throw new RuntimeException("Installing can be done only on startup")
     }
+
   }
 
+  def install(channels: Seq[C], patterns: Seq[P], continuation: K)(
```

Comment:
I'm not necessarily opposed to such a refactor, but if we decide to take this approach, let's break it out into a separate ticket and consistently, apply it to all the verbs.

Reply:
I agree, but without this I wouldn't be able to implement the new functionality